apps/emqx_dashboard/src/emqx_dashboard_login_lock.erl

```diff
@@ -0,0 +1,280 @@
+%% Copyright (c) 2025 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+-module(emqx_dashboard_login_lock).
+
+-include("emqx_dashboard.hrl").
+
+-include_lib("emqx/include/logger.hrl").
+
+-behaviour(gen_server).
+
+-export([
+    register_unsuccessful_login/1,
+    verify/1,
+    reset/1,
+    is_login_locked_error/1
+]).
+
+-export([
+    create_tables/0,
+    start_link/0
+]).
+
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    terminate/2
+]).
+
+%% For testing/debugging
+-export([
+    cleanup_all/0,
+    info/0
+]).
+
+-type time_s() :: non_neg_integer().
+-type time_us() :: non_neg_integer().
+-type username() :: binary().
+
+-record(login_lock, {
+    username :: username() | '_',
+    locked_until :: time_s() | '_'
+}).
+
+%% NOTE
+%% We ignore the fact that we may lose some attempts
+%% if they happen in the same microsecond.
+-record(login_attempts, {
+    key :: {username(), time_us()} | {'_', '_'} | '_',
+    extra :: map() | '_'
+}).
+
+-define(TAB_ATTEMPTS, login_attempts).
+-define(TAB_LOCK, login_lock).
+
+-ifdef(TEST).
+-define(CLEANUP_INTERVAL, 100).
+-else.
+-define(CLEANUP_INTERVAL, timer:minutes(30)).
+-endif.
+
+-define(ETS_CHUNK_SIZE, 100).
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+-spec register_unsuccessful_login(username()) -> ok.
+register_unsuccessful_login(Username) ->
+    NowUs = now_us(),
+    Key = {Username, NowUs},
+    Attempt = #login_attempts{key = Key, extra = #{}},
+    max_attempts_reached(Username, NowUs) andalso lock(Username, NowUs),
+    ok = mria:dirty_write(?TAB_ATTEMPTS, Attempt).
+
+-spec verify(username()) -> ok | {error, login_locked}.
+verify(Username) ->
+    Now = erlang:system_time(second),
+    case mnesia:dirty_read(?TAB_LOCK, Username) of
+        [#login_lock{locked_until = LockedUntil}] when LockedUntil > Now ->
+            {error, login_locked};
+        _ ->
+            ok
+    end.
+
+-spec is_login_locked_error(term()) -> boolean().
+is_login_locked_error(login_locked) ->
+    true;
+is_login_locked_error(_) ->
+    false.
+
+-spec reset(username()) -> ok.
+reset(Username) ->
+    cleanup(Username).
+
+-spec info() ->
+    #{
+        failed_attempt_records := non_neg_integer(),
+        lock_records := non_neg_integer()
+    }.
+info() ->
+    #{
+        failed_attempt_records => mnesia:table_info(?TAB_ATTEMPTS, size),
+        lock_records => mnesia:table_info(?TAB_LOCK, size)
+    }.
+
+%%--------------------------------------------------------------------
+%% Internal API
+%%--------------------------------------------------------------------
+
+create_tables() ->
+    ok = mria:create_table(?TAB_ATTEMPTS, [
+        {type, ordered_set},
+        {rlog_shard, ?DASHBOARD_SHARD},
+        {storage, disc_copies},
```

Comment:
Why disc_copies and not, say, rocksdb_copies? Do we need to cache contents of the entire table in RAM?

Reply:
The number of records is assumed to be the same order as the number of users, i.e. quite small.

Reply:
Do we need this data constantly in RAM? It looks like this table is read only after an unsucessful login attempt to calculate the number of entries in a certain interval. 

This event shouldn't happen at high frequency, so `mnesia:select` should do.